### PR TITLE
Requirement verification SDK logic #86

### PR DESCRIPTION
**Pull request by lxatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/CDP4-SDK-Community-Edition/pulls) open
- [x] I have verified that I am following the CDP4-SDK [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/CDP4-SDK-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description
- Implemented logic for calculation/verification of requirements
- Added unit test for the verification logic
- In the dec 11 commits also changes to project files regarding Nuget pre-release versions were made

In a nutshell:
Based on a Poco-tree:

RequirementSpecification
      - Requirement
            - ParametricConstraint
                   - (BooleanExpression tree)
                           - RelationalExpression

...a tree of Verifiers is created for verification.

All ValueSets are used, no matter what ParameterType and how many Options, or States are used for a single parameter:
All ValueSets/ValueArrays Pass? result = Pass
All ValueSets/ValueArrays Failed? result = Failed
In any other case: result = Inconclusive.

Future development, for example verification using an Option filter, or finetuning ValueSets based on different ActualFiniteStates is probably inevitable.

### CODE REVIEW

File: CDP4ServicesDal/CDP4ServicesDal.csproj [hunk 1]
```diff
@@ -4,7 +4,7 @@
     <TargetFrameworks>net45;net451;net452;net46;net461;net462;net47;net471;net472;netstandard2.0</TargetFrameworks>
     <Company>RHEA System S.A.</Company>
     <Title>CDP4ServicesDal Community Edition</Title>
-    <Version>5.0.5</Version>
+    <VersionPrefix>5.0.6</VersionPrefix>
```
Comment (samatrhea) [inline CDP4ServicesDal/CDP4ServicesDal.csproj]: see previous comment and revert

File: CDP4Rules/CDP4Rules.csproj [hunk 2]
```diff
@@ -4,7 +4,7 @@
     <TargetFrameworks>net45;net451;net452;net46;net461;net462;net47;net471;net472;netstandard2.0</TargetFrameworks>
     <Company>RHEA System S.A.</Company>
     <Title>CDP4Rules Community Edition</Title>
-    <Version>1.2.0</Version>
+    <VersionPrefix>1.2.1</VersionPrefix>
```
Comment (samatrhea) [inline CDP4Rules/CDP4Rules.csproj]: see previous comment and revert

File: CDP4Common/CDP4Common.csproj [hunk 3]
```diff
@@ -4,7 +4,7 @@
     <TargetFrameworks>net45;net451;net452;net46;net461;net462;net47;net471;net472;netstandard1.6;netstandard2.0</TargetFrameworks>
     <Company>RHEA System S.A.</Company>
     <Title>CDP4Common Community Edition</Title>
-    <Version>4.0.0</Version>
+    <VersionPrefix>4.0.1</VersionPrefix>
```
[the remark this answers is not in the document]
Comment (lxatrhea) [reply CDP4Common/CDP4Common.csproj]: Because we create pre-release on DevOps that need a "higher" version than the last one.
That is done by using a prefix and a suffix part to construct the version number.
In the "nuget package world" 4.0.0-prefix is a lower version than 4.0.0.
4.0.0 is a release, so for the next pre-release the version has to be higher.
I choose to add a +1 to the patch number. 
Indeed we can update the minor, or even the major version number when we create a release.

File: CDP4Requirements/ExtensionMethods/RequirementStateOfComplianceMethods.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+// -----------------------------------------------------------------------------------------------
+
+namespace CDP4Requirements.ExtensionMethods
+{
+    using System.Linq;
+
+    /// <summary>
+    /// Extension methods for the <see cref="RequirementStateOfCompliance"/> Enum
+    /// </summary>
+    public static class RequirementStateOfComplianceMethods
```
[the remark this answers is not in the document]
Comment (lxatrhea) [reply CDP4Requirements/ExtensionMethods/RequirementStateOfComplianceMethods.cs]: RequirementStateOfCompliance in an Enum. You can't add methods directly to Enums.

File: CDP4Requirements/Calculations/RequirementStateOfComplianceCalculator.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+                    return comparer.Compare(valueSet.ActualValue, relationalExpression.Value) != 0
+                        ? RequirementStateOfCompliance.Pass
+                        : RequirementStateOfCompliance.Failed;
+
+                default: throw new ArgumentOutOfRangeException(nameof(relationalExpression.RelationalOperator), relationalExpression.RelationalOperator, $"Unknown {nameof(relationalExpression.RelationalOperator)}");
+            }
+        }
+
+        /// <summary>
+        /// Gets the ValueSet comparer for this kind of <see cref="ScalarParameterType"/>
```
[the remark this answers is not in the document]
Comment (lxatrhea) [reply CDP4Requirements/Calculations/RequirementStateOfComplianceCalculator.cs]: No, we do, but only at a string comparison level. So a complete ValueArray<string> is compared.
Finetuning/follow up is open for discussion in https://github.com/RHEAGROUP/CDP4-IME-Community-Edition/issues/253
